[2.19] Add content-hashing to pub docs (#4407)
dart-lang/site-www: src/tools/pub/glossary.md

[lockfiles]: #lockfile
[version constraints]: #version-constraint

## Content hashes

> The first time `dart pub get` runs,
> it computes a sha256 hash for each hosted package,
> derived from the contents of their [lockfiles][].
> The hash embeds in the description of `pubspec.lock`.
> The pub client uses content hashes to
> verify the integrity of downloaded archives
> and detect if something has changed on the server.
> On future `pub get` runs, 
> the client can verify descriptions provided by the server
> against the descriptions of its cached packages.
> If they don't match, the archive is redownlaoded.
> If they still don't match, the resolution fails with an error.
> 
> To utilize this feature during package resolution,
> call [`dart pub get --enforce-lockfile`][].
> 
> [lockfiles]: #lockfile
> [`dart pub get --enforce-lockfile`]: /tools/pub/cmd/pub-get#--enforce-lockfile

Reviewer: ```suggestion
The pub.dev repository serves the sha256 hash of each package version it hosts.
This is useful for the pub client to validate the integrity of the downloaded package. 

When `dart pub get` downloads a package, it computes the hash of the downloaded archive and re-downloads if there is a discrepancy. 

But furthermore the hash is used to protect against changes on the repository.

The hash of each hosted dependency will be stored with the resolution in the [lockfile][].

The pub client uses this content hash to verify that exactly the same packages are used when running `pub get` again using the same lock file, potentially on a different computer.

If the locked hash doesn't match what's currently in the pub cache, the archive is redownloaded.
If it still doesn't match, the lockfile is updated and a warning is printed like:
```
$ dart pub get
Resolving dependencies...
Cached version of foo-1.0.0 has wrong hash - redownloading.
 ~ foo 1.0.0 (was 1.0.0)
The existing content-hash from pubspec.lock doesn't match contents for:
 * foo-1.0.0 from "pub.dev"
This indicates one of:
 * The content has changed on the server since you created the pubspec.lock.
 * The pubspec.lock has been corrupted.
 
The content-hashes in pubspec.lock has been updated.

For more information see:
https://dart.dev/go/content-hashes

Changed 1 dependency!
```

This updated hash will now show up in your version control diff and should make you suspicious.

To make a discrepancy become an error instead of a warning use [`dart pub get --enforce-lockfile`][] that will fail if it cannot find package archives with the same hashes without updating the lockfile.

```
$ dart pub get --enforce-lockfile
Resolving dependencies...
Cached version of foo-1.0.0 has wrong hash - redownloading.
~ foo 1.0.0 (was 1.0.0)
The existing content-hash from pubspec.lock doesn't match contents for:
 * foo-1.0.0 from "pub.dev"

This indicates one of:
 * The content has changed on the server since you created the pubspec.lock.
 * The pubspec.lock has been corrupted.

For more information see:
https://dart.dev/go/content-hashes
Would change 1 dependency.
Unable to satisfy `pubspec.yaml` using `pubspec.lock`.

To update `pubspec.lock` run `dart pub get` without
`--enforce-lockfile`.
```

[lockfile]: #lockfile
[`dart pub get --enforce-lockfile`]: /tools/pub/cmd/pub-get#--enforce-lockfile
```

Author: This is excellent @sigurdm thanks so much for writing it up!

[lockfiles]: #lockfile
[version constraints]: #version-constraint

## Content hashes

> The pub.dev repository serves a sha256 hash of each package version it hosts.

Reviewer 2: **question:** "Serves" reads odd here. How is a hash served? Is it listed with the package? Is it included as part of the download? Also, looking to the next sentence, make it clear that `pub` is doing this validation, that is, The `pub` client validates a package's integrity using this hash.

Author: >  Also, looking to the next sentence, make it clear that pub is doing this validation, that is, The pub client validates a package's integrity using this hash.

The second sentence already says literally that exact same thing: 

> "The pub client can use this hash to validate the integrity of downloaded packages,
> and protect against changes on the repository." 

So I'm not sure what you're asking for there.

The pub client can use this hash to validate the integrity of downloaded packages,
and protect against changes on the repository. 

When `dart pub get` downloads a package, it computes the hash of the
downloaded archive. The hash of each hosted dependency is stored with the
[resolution][] in the [lockfile][].

The pub client uses this content hash to verify that running `pub get` again using

> the same lockfile, potentially on a different computer, uses exactly the same packages.

Reviewer 3: Maybe "exactly the same package" could be made a bit more specific, since it could protect from matching packages, just modified or outdated.

I know pub uses the term ["package archive"](https://github.com/dart-lang/pub/blob/master/doc/repository-spec-v2.md#list-all-versions-of-a-package) to describe the saved contents of the file, but perhaps that would sound confusing here. Any thoughts?

I'll think about it some more, but if no one has ideas, I suppose we can just leave this as is.

Author: Maybe "uses the same packages in the same state", or "uses the correct package states"?

I don't really understand the topic enough to give a better suggestion

If the locked hash doesn't match what's currently in the pub cache,
pub redownloads the archive. If it still doesn't match, the lockfile
updates and a warning is printed. For example:

> ```

Author: Good idea!

$ dart pub get
Resolving dependencies...
Cached version of foo-1.0.0 has wrong hash - redownloading.
 ~ foo 1.0.0 (was 1.0.0)
The existing content-hash from pubspec.lock doesn't match contents for:
 * foo-1.0.0 from "pub.dev"
This indicates one of:
 * The content has changed on the server since you created the pubspec.lock.
 * The pubspec.lock has been corrupted.
 
The content-hashes in pubspec.lock has been updated.

For more information see:
https://dart.dev/go/content-hashes

Changed 1 dependency!
```

The updated content hash will show up in your version control diff,
and should make you suspicious.

To make a discrepancy become an error instead of a warning, use
[`dart pub get --enforce-lockfile`][]. It will cause the resolution to fail

> if it cannot find package archives with the same hashes, without updating the lockfile.

Reviewer 2: **issue:** Include some version of this in the explanation of `--enforce-lockfile`.

Author: Which part exactly? It seems to me to say the exact same thing: "...will refuse to resolve
if anything deviates from `pubspec.lock`"

Reviewer 2: This copy block provides context to resolution: package hashes don't match and you can't update the lock file. This is a great explanation.

relies on.

Unlike the pubspec, which only lists immediate dependencies and allows version
ranges, the lockfile comprehensively pins down the entire dependency graph to
specific versions of packages. A lockfile ensures that you can recreate the
exact configuration of packages used by an application.

The lockfile is generated automatically for you by pub when you run
[`pub get`](/tools/pub/cmd/pub-get), [`pub upgrade`](/tools/pub/cmd/pub-upgrade),

> or [`pub downgrade`](/tools/pub/cmd/pub-downgrade). Pub also computes a [content
> hash][] to check against future resolutions.

Reviewer 3: Perhaps this is less of an "also" and more of a first class functionality:

```suggestion
or [`pub downgrade`](/tools/pub/cmd/pub-downgrade). 
Pub includes a [content hash][] for each package
to check against during future resolutions.
```

relies on.

Unlike the pubspec, which only lists immediate dependencies and allows version
ranges, the lockfile comprehensively pins down the entire dependency graph to
specific versions of packages. A lockfile ensures that you can recreate the
exact configuration of packages used by an application.

The lockfile is generated automatically for you by pub when you run
[`pub get`](/tools/pub/cmd/pub-get), [`pub upgrade`](/tools/pub/cmd/pub-upgrade),
or [`pub downgrade`](/tools/pub/cmd/pub-downgrade).

> Pub inlcudes a [content hash][] for each package

Reviewer 3: ```suggestion
Pub includes a [content hash][] for each package
```
